Adding support for upper and lower bound functions in SSA

Gamrix opened the pull request:
This diff adds support for `torch.nonzero` as a torch function that has both an upper and a lower bound size for shape functions, and therefore generates new symbolic dimensions in JIT.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77389
* #77388

Gamrix committed:
Adding support for upper and lower bound functions in SSA

[ghstack-poisoned]

Gamrix committed:
Update on "Adding support for upper and lower bound functions in SSA"




[ghstack-poisoned]

Gamrix committed:
Update on "Adding support for upper and lower bound functions in SSA"




[ghstack-poisoned]

Gamrix commented:
@miladm Adding support for is_dynamic for torch.nonzero .

Gamrix committed:
Update on "Adding support for upper and lower bound functions in SSA"

This diff adds support for `torch.nonzero` as a torch function that has both an upper and a lower bound size for shape functions, and therefore generates new symbolic dimensions in JIT.




[ghstack-poisoned]

Gamrix committed:
Update on "Adding support for upper and lower bound functions in SSA"

This diff adds support for `torch.nonzero` as a torch function that has both an upper and a lower bound size for shape functions, and therefore generates new symbolic dimensions in JIT.




[ghstack-poisoned]

eellison reviewed (commented):
Looking good! A few comments

Gamrix commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -237,34 +242,58 @@ void transformShapeFunction(
     const std::string& shape_compute_function_name,
     std::unordered_map<std::string, std::shared_ptr<Graph>>& reused_functions,
     const CompilationUnit& module) {
+  // TODO: When elias edits transformShapeFunction, remove the unnecessary
```
Thought this diff would remove the need for the schema string, but clearly Ia m wrong.

https://github.com/pytorch/pytorch/pull/77120/files

Gamrix commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -747,6 +754,24 @@ std::shared_ptr<Graph> PropagateShapesWithShapeFunction(
+c10::SymbolicShape combine_bounds(
+    c10::SymbolicShape& lower_bound,
+    c10::SymbolicShape& upper_bound) {
+  TORCH_INTERNAL_ASSERT(lower_bound.rank() == upper_bound.rank());
```
I'll add a TODO for that. I don't think there are many cases in SSA where we will have different upper and lower number of dimmensions.

Gamrix commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -237,34 +242,58 @@ void transformShapeFunction(
+    const std::string& upper_bound_function_name,
+    std::unordered_map<std::string, std::shared_ptr<Graph>>& reused_functions,
+    const CompilationUnit& module) {
+  auto lower_graph = addFunctionToCache(
```
I renamed the function to reduce confusion.

Gamrix commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -237,34 +242,58 @@ void transformShapeFunction(
+    const std::string& shape_compute_function_name,
+    std::unordered_map<std::string, std::shared_ptr<Graph>>& reused_functions,
+    const CompilationUnit& module) {
+  auto graph = addFunctionToCache(
```
I renamed the function to reduce confusion.

Gamrix committed:
Update on "Adding support for upper and lower bound functions in SSA"

This diff adds support for `torch.nonzero` as a torch function that has both an upper and a lower bound size for shape functions, and therefore generates new symbolic dimensions in JIT.




[ghstack-poisoned]

Gamrix commented:
Updated to address comments

Gamrix commented:
@pytorchmergebot merge this please
